File: faultbench/micro_replays/python_requests_7308/replay.py
#!/usr/bin/env python3
r"""Standalone semantic replay of Requests PR 7308's `$` to `\Z` regex change."""

from __future__ import annotations

import json
import re


def main() -> None:
    before = re.compile(r"^\S[^\r\n]*$|^$")
    after = re.compile(r"^\S[^\r\n]*\Z|^\Z")
    invalid = "bar\n"
    before_incorrectly_accepts = before.match(invalid) is not None
    after_rejects = after.match(invalid) is None
    valid_still_accepts = after.match("bar") is not None
    assert before_incorrectly_accepts
    assert after_rejects
    assert valid_still_accepts
    print(
        json.dumps(
            {
                "case_id": "requests-7308",
                "language": "Python",
                "before_defect_observed": before_incorrectly_accepts,
                "after_expectation_passed": after_rejects and valid_still_accepts,
                "scope": "standalone semantic micro-replay; not the Requests test suite",
            },
            sort_keys=True,
        )
    )


if __name__ == "__main__":
    main()

File: faultbench/micro_replays/javascript_fastify_6838/replay.js
'use strict'

// Standalone semantic replay of Fastify PR 6838's method normalization change.
const assert = require('node:assert/strict')

const routes = new Map([['GET /artists/:artistId', { params: { artistId: ':artistId' } }]])

function beforeFindRoute (options) {
  return routes.get(`${options.method} ${options.url || ''}`) ?? null
}

function afterFindRoute (options) {
  return routes.get(`${options.method?.toUpperCase() ?? ''} ${options.url || ''}`) ?? null
}

const request = { method: 'get', url: '/artists/:artistId' }
const beforeDefectObserved = beforeFindRoute(request) === null
const afterResult = afterFindRoute(request)
const afterExpectationPassed = afterResult?.params.artistId === ':artistId'
const missingMethodStillNull = afterFindRoute({ url: request.url }) === null

assert.equal(beforeDefectObserved, true)
assert.equal(afterExpectationPassed, true)
assert.equal(missingMethodStillNull, true)

process.stdout.write(JSON.stringify({
  case_id: 'fastify-6838',
  language: 'JavaScript',
  before_defect_observed: beforeDefectObserved,
  after_expectation_passed: afterExpectationPassed && missingMethodStillNull,
  scope: 'standalone semantic micro-replay; not the Fastify test suite'
}) + '\n')
